src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -185,25 +199,50 @@ def providers_for_module(self, module: str) -> tuple[ModuleProvider, ...]:
         if "." not in module:
             return ()
         parent_module = module.rsplit(".", maxsplit=1)[0]
-        return self.providers_for_module(parent_module)
+        return self._providers_for_resolve(parent_module, resolve)
+
+    def providers_for_module(
+        self, module: str, resolves: Iterable[str] | None
+    ) -> tuple[ModuleProvider, ...]:
+        """Find all providers for the module.
+
+        If `resolves` is None, will not consider resolves, i.e. any `python_requirement` can be
+        consumed. Otherwise, providers can only come from `python_requirements` marked compatible
+        with those resolves.
+        """
+        if resolves is None:
+            resolves = list(self.keys())
```

Comment:
We'll continue to set `resolves=None` if `--no-python-enable-resolves` (the default).